02-peer-protocol.md

```diff
@@ -787,9 +787,21 @@ is destined, is described in [BOLT #4](04-onion-routing.md).
 #### Requirements
 
 A sending node:
-  - MUST NOT offer `amount_msat` it cannot pay for in the
-remote commitment transaction at the current `feerate_per_kw` (see "Updating
-Fees") while maintaining its channel reserve.
+  - if it is _responsible_ for paying the Bitcoin fee:
+    - MUST NOT offer `amount_msat` if, after adding that HTLC to its commitment
+    transaction, it cannot pay the fee for either the local or remote commitment
+    transaction at the current `feerate_per_kw` while maintaining its channel
+    reserve (see [Updating Fees](#updating-fees-update_fee)).
+    - SHOULD NOT offer `amount_msat` if, after adding that HTLC to its commitment
+    transaction, its remaining balance doesn't allow it to pay the fee for a
+    future additional non-dust HTLC at a higher feerate while maintaining its
+    channel reserve ("fee spike buffer"). A buffer of `2*feerate_per_kw` is
+    recommended to ensure predictability.
+  - if it is _not responsible_ for paying the Bitcoin fee:
+    - MUST NOT offer `amount_msat` if, once the remote node adds that HTLC to
+    its commitment transaction, it cannot pay the fee for the updated local or
+    remote transaction at the current `feerate_per_kw` while maintaining its
+    channel reserve.
```

Comment:
So is the sender supposed to check whether the remote funder can pay on-chain fees for the receiver's **next** commitment tx at corresponding feerate1, and also whether the remote funder can pay on-chain fees for the sender's **next** commitment tx at corresponding feerate2? (note that e.g. dust limits might differ so both ctxs need to be inspected)

If so, what is feerate2? The sender (being the fundee) has no way of knowing the actual future feerate for that tx at this point.

Indeed while the sender is running these checks and considering sending an HTLC, if the remote funder sends an update_fee exactly at that time the actual feerate2 value will change.
Note that the feerate changing implies the htlc trim threshold changing, so even dust value HTLCs are not safe to send due to this race (as they might not be dust at that later time).

Reply:
> If so, what is feerate2? The sender (being the fundee) has no way of knowing the actual future feerate for that tx at this point.

Both peers know the feerate. It's what's currently applied to the commit tx. There is no "feerate2" here, the only thing both know is the current feerate.

Obviously there may be concurrent `update_fee`, but you can't predict those. Depending on what is signed and revoked first (the HTLC or an `update_fee`) one of the two may fail. The faster will succeed, the slower will fail.

Reply:
> Depending on what is signed and revoked first (the HTLC or an update_fee) one of the two may fail. The faster will succeed, the slower will fail.

Sure but it might fail in a way that such there will be a commitment tx with the channel reserve violated. For example, local add htlc and commits right away, while remote concurrently updates (increases) fee and commits right away. Now, in the "next" commitment tx of remote, they might not afford the onchain fees (they could have afforded the fees if either one of the updates did not occur).
As remote dipped below channel reserve, and the updates are now committed, depending on implementation, the channel might get force-closed at that point.

Note that if there was also a "fee spike buffer" in this scenario, the fatal race condition would become less likely.

Reply:
No, we don't force close when the remote violates this additional margin. Its a choice how much margin the remote calculates though.   2*feerate is just the default suggestion.

Reply:
The different interpretations will just lead to `spendable_msat` and `receivable_msat` calculations being wrong, leading to potentially CAPACITY EXCEEDED errors when someone tries to drain a channel.

Reply:
> Sure but it might fail in a way that such there will be a commitment tx with the channel reserve violated.

That concurrency has always existed. It's mentioned in the `Updating Fees` paragraph below. We will continue handling those races as detailed in that paragraph. But the outcome isn't a force-close; on the contrary, the outcome is that the current commitment tx may have a lower feerate than what the funder wishes it to be because it couldn't go through with his `update_fee`.

Reply:
> Both peers know the feerate. It's what's currently applied to the commit tx. There is no "feerate2" here, the only thing both know is the current feerate.

Also, there are actually two different feerates, if e.g.:
- Remote sends update_fee (changes fee from feerate1 to feerate2), and local receives this. It is not yet committed.
- Local wants to send update_add_htlc. Now, what feerates should local use when running the checks? The next remote ctx will still use feerate1, while the next local ctx will use feerate2.

Reply:
There is only one latest feerate that has been signed by both parties (everything else is just a pending change).

> Local wants to send update_add_htlc

If local chooses to apply and sign the `update_fee` with the new feerate, it uses the new feerate. Otherwise it uses the previous feerate. Choosing which one it applies first (or both at the same time) is an implementation choice and all are valid from the spec's point of view.

I feel this discussion is not related to stuck channels and the fee spike buffer though. If you want to further discuss `update_fee` and its asynchronous nature, can you please open an issue where we'd continue? I'd really like to make progress on stuck channels and I fear this side discussion is going to add more unrelated cognitive load on reviewers.

Reply:
> That concurrency has always existed. It's mentioned in the `Updating Fees` paragraph below. We will continue handling those races as detailed in that paragraph. 

Ah, you are right, I forgot about that. I guess that contradicts with this other paragraph though?

> The channel reserve is specified by the peer's channel_reserve_satoshis: 1% of the channel total is suggested. Each side of a channel maintains this reserve so it always has something to lose if it were to try to broadcast an old, revoked commitment transaction. Initially, this reserve may not be met, as only one side has funds; **but the protocol ensures that there is always progress toward meeting this reserve, and once met, it is maintained**.

If this race happens, the channel reserve can be violated.
And a malicious node can intentionally keep trying to trigger the race.

Reply:
> I feel this discussion is not related to stuck channels and the fee spike buffer though. If you want to further discuss update_fee and its asynchronous nature, can you please open an issue where we'd continue? 

This discussion might not be related to "stuck channels" but it is certainly related to the highlighted text at the top of this thread which is new text added by this PR.
I am questioning what current `feerate_per_kw` is.
In any case, as there is no corresponding check by the receiver, I guess it's fine if everyone interprets it their own way. -- it only really matters if you intend to add a corresponding receiver check later.

Reply:
> I guess that contradicts with this other paragraph though?
> If this race happens, the channel reserve can be violated.

No it cannot. I don't understand why you're claiming that...?

Let's ignore the initial period where one side has 0 funds, and jump to the moment where both sides have funds above the reserve. There is no situation where the channel reserve will be violated, and this PR doesn't change that invariant. I really don't see why you think it does.

Reply:
Well this part (channel reserve) of this conversation is indeed getting off-topic a bit, I tried to find you on IRC but could not.
In any case, as established above, a race can happen, where the funder cannot actually afford the onchain fees, and in that case this paragraph applies:
> Base commitment transaction fees are extracted from the funder's amount; if that amount is insufficient, the entire amount of the funder's output is used.

That is, it could happen that the funder does not have an output in the commitment tx. In the extreme case, due to the race, given appropriate dust_limit and feerate values, it could happen that the funder's output (would it not be pruned) would be 0 sat. If after fees the funder's output would be 0 sat, how could the channel reserve be met?

Reply:
> I'll open a new issue just for this channel reserve part.

Please do, that will help others join and help clarify this without side-tracking this PR.
This topic had already been discussed a lot in the past, it's worth doing a bit of research to dig up those discussions.